packages/user/src/User/Login.purs

```diff
@@ -198,38 +199,61 @@ failOnEmailMismatch self email
       liftEffect $ self.setState _ { errors { login = Just LoginEmailMismatchError } }
       throwError $ error "Emails don't match"
 
+loginWithRetry :: (Aff (Either UserError User)) -> (User -> Aff Unit) -> (UserError -> Aff Unit) -> (Either UserError User -> Aff Unit) -> Aff Unit
+loginWithRetry action handleSuccess handleError handleFinish = withRetry false
+  where
+    handleError' err = do
+      handleError err
+      pure true
+    withRetry retry = do
+      user <- action
+      notRetry <- case user of
+        Right user' -> do
+          handleSuccess user'
+          pure true
+        Left ServiceUnavailable -> do
+          if retry then handleError' ServiceUnavailable else do
+            delay $ Milliseconds 1000.0
+            withRetry true
+            pure false
+        Left err -> handleError' err
+      when notRetry $ handleFinish user
```

Comment:
I think this would be neater to read without `notRetry`, even though it becomes a bit repetitive:
```
    withRetry retry = do
      user <- action
      case user of
        Right user' -> do
          handleSuccess user'
          handleFinish user
        Left ServiceUnavailable -> do
          if retry
          then do
            handleError ServiceUnavailable
            handleFinish user
          else do
            delay $ Milliseconds 1000.0
            withRetry true
        Left err -> do
          handleError err
          handleFinish user
```

Reply:
I updated the function. I moved `handleFinish` from inside the `withRetry` to a bind immediately after calling `withRetry` on the main `loginWithRetry` function. I think this emphasizes it that no matter which way you run the function, that's what gets used last.